src/hevm/src/EVM/Symbolic.hs

```diff
@@ -147,9 +151,14 @@ select' xs err ind = walk xs ind err
 -- the index is symbolic, but it still seems (kind of) manageable
 -- for the solvers.
 readSWordWithBound :: SWord 32 -> Buffer -> SWord 32 -> SymWord
-readSWordWithBound ind (SymbolicBuffer xs) bound =
-  let boundedList = [ite (i .<= bound) x 0 | (x, i) <- zip xs [1..]]
-  in sw256 . fromBytes $ [select' boundedList 0 (ind + j) | j <- [0..31]]
+readSWordWithBound ind (SymbolicBuffer xs) bound = case (num <$> fromSized <$> unliteral ind, num <$> fromSized <$> unliteral bound) of
+  (Just i, Just b) ->
+    let bs = truncpad 32 $ drop i (take b xs)
+    in S (FromBytes (SymbolicBuffer bs)) (fromBytes bs)
+  _ -> 
+    let boundedList = [ite (i .<= bound) x 0 | (x, i) <- zip xs [1..]]
+    in sw256 . fromBytes $ [select' boundedList 0 (ind + j) | j <- [0..31]]
```

Comment:
can't really follow this, what is `select'` doing?

Reply:
specialized to `select' :: [SWord 8] -> SWord 8 -> Int -> SWord 8` here, `select'` creates a massive folded `ite` over all of the elements of the list, which picks the element of the list at index `(ind + j)` if in range, otherwise 0.

Reply:
n.b. that this is written using list comprehension syntax https://wiki.haskell.org/List_comprehension

Reply:
ok got it now, it builds a new 32 element list, where each element is some giant hairy `ite` over the original list selecting the element at `ind + j`.

Why do we need the `select'` construct? Why not use `readArray boundedList (ind + j)`?

Reply:
Because static calladata are modeled as a (haskell) list, not an array

Reply:
ah yes, of course....